packages/eslint-plugin/src/rules/no-confusing-void-expression.ts

```diff
@@ -376,5 +427,55 @@ export default createRule<Options, MessageId>({
       const type = getConstrainedTypeAtLocation(services, targetNode);
       return tsutils.isTypeFlagSet(type, ts.TypeFlags.VoidLike);
     }
+
+    function hasValidReturnType(
+      services: ParserServicesWithTypeInformation,
+      node:
+        | TSESTree.FunctionExpression
+        | TSESTree.ArrowFunctionExpression
+        | TSESTree.FunctionDeclaration
+        | null,
+    ): boolean {
+      if (node != null) {
+        const checker = services.program.getTypeChecker();
+        const functionTSNode = services.esTreeNodeToTSNodeMap.get(node);
+        const functionType =
+          (ts.isFunctionExpression(functionTSNode) ||
+          ts.isArrowFunction(functionTSNode)
+            ? getContextualType(checker, functionTSNode)
+            : services.getTypeAtLocation(node)) ??
+          services.getTypeAtLocation(node);
+
+        if (functionTSNode.type) {
+          for (const signature of tsutils.getCallSignaturesOfType(
+            functionType,
+          )) {
+            return !(
```

Comment:
[Bug] This code returns from `for` loop on the first iteration. So if a function has several signatures, the rule will check only the first signature:

```ts
function test1(): void
function test1(arg: string): any
function test1(arg?: string): any | void {
  if (arg) {
    return arg
  }

  return console.log() // not reported - ok
}

function test2(arg: string): any
function test2(): void
function test2(arg?: string): any | void {
  if (arg) {
    return arg
  }

  return console.log() // reported - bug
}
```

[playground](https://deploy-preview-8632--typescript-eslint.netlify.app/play/#ts=5.4.3&fileType=.ts&code=GYVwdgxgLglg9mABFApgZygRgBQEoBciAbnDACYBQoksCy6W2AhgE4DmhGLMYbBiTMAE8q4aPCSoMOVmwD8nKN179BQxAB9ipMogDeFRIhjBEzdrn2GjiFiiggWSWdYC%2BFa3YdPEEBGjgAGxQAOkC4NjxEAHpoxDA4KFsUAAc4FlRdAFpEOABrCndRGgl6DAAmcw5ELh4%2BQjVi8TopKEr%2BEnIm2kkGStkFGqU61WFNbXIrIxMzWUsDG2TvZ3Y3DyMvRyQ-MADgsIio2OS0jJRsxAAjEDZCoA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1rI6YDNYyZgHNaANw6UAJvQAexaCiG90YANqRE0aB2iQANJkiVhnRQDVJUgJJNL01fjiIAvgF1wYFyBdA&tsconfig=&tokens=false)